### DISCUSSION TITLE
Observability: OpenTelemetry

### DISCUSSION THREAD

**Discussion by snovak7** [keycloak/keycloak#28369] [Ideas]
Don't know how to begin, but the idea to have a consolidated Observability with Open Telemetry...
After researching, the metrics & tracing are part of Quarkus (though in limited format at the moment), but was thinking for Logs if Log Handler would need to be implemented to send logs to Open Telemetry if it is defined in ENV (also crucial for Metrics, Tracing) as part of instrumentation?

**Comment by codespearhead** [comment]
I'm not sure I understand the question, but does [this](https://www.keycloak.org/keycloak-benchmark/kubernetes-guide/latest/util/otel) answer your question?

**Comment by snovak7** [comment]
Hello @codespearhead

Thank you for reply,

I've seen this article before, and this is how I concluded first part of the narrative. Since I don't come from Java background (was in the past), and am thinking about instead of reading log files on system/pod level, I would rather send them in context of traces (+ structured logging), not sure if that (structured logging) is a thing in java space.

For now I am interested if log handler could make this happen...

Though my ultimate goal/wish would be that OTEL would be primary citizen in observability or as an option

I hope I am clearer in explaining the story here

**Reply by codespearhead** [reply]
I'm afraid the question has more to with Prometheus (or Granafa Loki) + Grafana than with Keycloak.

OTEL enables you to export observability data to Prometheus, so you can create visualizations with Grafana.

**Reply by greyarch** [reply]
I think @snovak7's problem is that OTEL, as implemented in keycloak, does not deal with logs at all, only metrics and traces. There is nothing in the documentation about OTEL logs setup.
I have exactly the same problem.

**Comment by greyarch** [comment]
Hey @snovak7, I think I have the same issue and would be very interested to know if you found a solution?
In short, I want to send Keycloak logs to my OpenTelemtry collector. However, the documentation seems to only be concerned with traces, when it comes to OTEL. I, just like you, want to do everything with OTEL - metrics, traces and logs.

**Reply by snovak7** [reply]
Unfortunately no, I haven't invested into making it possible.

Would be killer feature that when enabling `KC_OTEL=true` would give you automatic instrumentation though given it finds Open Telemetry environmental variables of course.

**Reply by greyarch** [reply]
KC_OTEL=true should not even be required - OTEL feature is on by default now.

**Reply by abhiaagarwal** [reply]
@greyarch I've achieved okay results by using the rsyslog receiver in opentel and sending keycloak logs there, but a native solution should be possible.

**Reply by greyarch** [reply]
What I ended up doing is using syslog to send logs to my OTEL collector (the contrib image has it) and then sending logs over to splunk with the splunk_hec exporter that the collector supports.
I really like this solution because it gives me the ability to process (filter) logs before sending them anywhere. I can also send them to different storages, I can debug and do all sorts of things, the would otherwise be impossible or rather more difficult.

**Reply by RadekSalay** [reply]
Thank you @abhiaagarwal and @greyarch for direction on syslog, for sending logs from keycloak with native telemetry.

**Reply by mabartos** [reply]
OTel Logs and OTel Metrics are not supported right now by Keycloak, but we're going the path OTel-everywhere :P It just needs some refinements from Quarkus team as well.

**Reply by abhiaagarwal** [reply]
Thanks for the response @mabartos, very excited :)
